File: react-native-nitro-player/ios/core/TrackPlayerQueue.swift
//
//  TrackPlayerQueue.swift
//  NitroPlayer
//
//  Created by Ritesh Shukla on 25/03/26.
//

import AVFoundation
import Foundation

extension TrackPlayerCore {

  func getState() async -> PlayerState {
    await withPlayerQueueNoThrow { self.getStateInternal() }
  }

  func getActualQueue() async -> [TrackItem] {
    await withPlayerQueueNoThrow { self.getActualQueueInternal() }
  }

  @discardableResult
  func skipToIndexOnQueue(index: Int) -> Bool {
    isCasting ? skipToIndexCastInternal(index: index) : skipToIndexInternal(index: index)
  }

  func skipToIndex(index: Int) async -> Bool {
    await withPlayerQueueNoThrow { self.skipToIndexOnQueue(index: index) }
  }

  func playFromIndex(index: Int) async {
    await withPlayerQueueNoThrow { self.playFromIndexInternal(index: index) }
  }

  func getCurrentTrackIndex() async -> Int {
    await withPlayerQueueNoThrow { self.currentTrackIndex }
  }

  // MARK: - Internal (run on playerQueue)

  func getStateInternal() -> PlayerState {
    // While casting, report the receiver's cached state — the local AVQueuePlayer is paused/stale.
    if isCasting, let cast = castManager {
      let currentTrack = getCurrentTrack()
      let playingType: CurrentPlayingType
      if currentTrack == nil { playingType = .notPlaying }
      else {
        switch currentTemporaryType {
        case .none: playingType = .playlist
        case .playNext: playingType = .playNext
        case .upNext: playingType = .upNext
        }
      }
      return PlayerState(
        currentTrack: currentTrack.map { Variant_NullType_TrackItem.second($0) },
        currentPosition: cast.lastKnownRemotePosition,
        totalDuration: cast.lastKnownRemoteDuration,
        currentState: cast.cachedPlaybackState,
        currentPlaylistId: currentPlaylistId.map { Variant_NullType_String.second($0) },
        currentIndex: currentTrackIndex >= 0 ? Double(currentTrackIndex) : -1.0,
        currentPlayingType: playingType
      )
    }

    guard let player else {
      return PlayerState(
        currentTrack: nil, currentPosition: 0.0, totalDuration: 0.0,
        currentState: .stopped,
        currentPlaylistId: currentPlaylistId.map { Variant_NullType_String.second($0) },
        currentIndex: -1.0, currentPlayingType: .notPlaying
      )
    }
    let currentTrack = getCurrentTrack()
    let currentPosition = player.currentTime().seconds
    let rawDuration = player.currentItem?.duration.seconds ?? 0.0
    let totalDuration = (rawDuration > 0 && !rawDuration.isNaN && !rawDuration.isInfinite) ? rawDuration : 0.0

    let state: TrackPlayerState
    if player.timeControlStatus == .playing { state = .playing }
    else if player.timeControlStatus == .waitingToPlayAtSpecifiedRate { state = .buffering }
    else if player.rate == 0 { state = .paused }
    else { state = .stopped }

    let currentIndex: Double = currentTrackIndex >= 0 ? Double(currentTrackIndex) : -1.0

    let playingType: CurrentPlayingType
    if currentTrack == nil { playingType = .notPlaying }
    else {
      switch currentTemporaryType {
      case .none: playingType = .playlist
      case .playNext: playingType = .playNext
      case .upNext: playingType = .upNext
      }
    }

    return PlayerState(
      currentTrack: currentTrack.map { Variant_NullType_TrackItem.second($0) },
      currentPosition: currentPosition,
      totalDuration: totalDuration,
      currentState: state,
      currentPlaylistId: currentPlaylistId.map { Variant_NullType_String.second($0) },
      currentIndex: currentIndex,
      currentPlayingType: playingType
    )
  }

  func getActualQueueInternal() -> [TrackItem] {
    var queue: [TrackItem] = []
    queue.reserveCapacity(currentTracks.count + playNextStack.count + upNextQueue.count)

    // Add tracks before current (original playlist)
    // When a temp track is playing, include the original track at currentTrackIndex
    let beforeEnd = currentTemporaryType != .none
      ? min(currentTrackIndex + 1, currentTracks.count) : currentTrackIndex
    if beforeEnd > 0 { queue.append(contentsOf: currentTracks[0..<beforeEnd]) }

    // Add current track (temp or original)
    if let current = getCurrentTrack() { queue.append(current) }

    // Add playNext stack — skip the currently playing track by ID (already added as current)
    let currentId = activeCurrentTrackId
    if currentTemporaryType == .playNext, let currentId = currentId {
      var skipped = false
      for track in playNextStack {
        if !skipped && track.id == currentId { skipped = true; continue }
        queue.append(track)
      }
    } else if currentTemporaryType != .playNext {
      queue.append(contentsOf: playNextStack)
    }

    // Add upNext queue — skip the currently playing track by ID (already added as current)
    if currentTemporaryType == .upNext, let currentId = currentId {
      var skipped = false
      for track in upNextQueue {
        if !skipped && track.id == currentId { skipped = true; continue }
        queue.append(track)
      }
    } else if currentTemporaryType != .upNext {
      queue.append(contentsOf: upNextQueue)
    }

    // Add remaining original tracks
    if currentTrackIndex + 1 < currentTracks.count {
      queue.append(contentsOf: currentTracks[(currentTrackIndex + 1)...])
    }
    return queue
  }

  /// Length of the logical queue and the current track's position in it, derived
  /// arithmetically. `getActualQueueInternal()` copies every TrackItem to answer the
  /// same two questions — too expensive to run on every playback-state change.
  func actualQueueMetrics() -> (count: Int, position: Int) {
    let before: Int
    if currentTemporaryType != .none {
      before = min(currentTrackIndex + 1, currentTracks.count)
    } else {
      before = max(0, currentTrackIndex)
    }
    let hasCurrent = getCurrentTrack() != nil
    let playNextCount =
      currentTemporaryType == .playNext ? max(0, playNextStack.count - 1) : playNextStack.count
    let upNextCount =
      currentTemporaryType == .upNext ? max(0, upNextQueue.count - 1) : upNextQueue.count
    let after =
      currentTrackIndex + 1 < currentTracks.count ? currentTracks.count - (currentTrackIndex + 1) : 0

    let count = before + (hasCurrent ? 1 : 0) + playNextCount + upNextCount + after
    return (count, hasCurrent ? before : -1)
  }

  func getCurrentTrack() -> TrackItem? {
    if currentTemporaryType != .none, let trackId = activeCurrentTrackId {
      if currentTemporaryType == .playNext { return playNextStack.first(where: { $0.id == trackId }) }
      if currentTemporaryType == .upNext { return upNextQueue.first(where: { $0.id == trackId }) }
    }
    guard currentTrackIndex >= 0 && currentTrackIndex < currentTracks.count else { return nil }
    return currentTracks[currentTrackIndex]
  }

  @discardableResult
  func skipToIndexInternal(index: Int) -> Bool {
    let actualQueue = getActualQueueInternal()
    guard index >= 0 && index < actualQueue.count else { return false }

    // Calculate queue section boundaries using effective sizes
    // (reduced by 1 when current track is from that temp list)
    let currentPos = currentTemporaryType != .none
      ? currentTrackIndex + 1 : currentTrackIndex
    let effectivePlayNextSize = currentTemporaryType == .playNext
      ? max(0, playNextStack.count - 1) : playNextStack.count
    let effectiveUpNextSize = currentTemporaryType == .upNext
      ? max(0, upNextQueue.count - 1) : upNextQueue.count

    let playNextStart = currentPos + 1
    let playNextEnd = playNextStart + effectivePlayNextSize
    let upNextStart = playNextEnd
    let upNextEnd = upNextStart + effectiveUpNextSize
    let originalRemainingStart = upNextEnd

    // Case 1: Target is before current - rebuild from that playlist index
    if index < currentPos {
      _ = rebuildQueueFromPlaylistIndex(index: index)
      return true
    }

    // Case 2: Target is current - seek to beginning
    if index == currentPos {
      player?.seek(to: .zero)
      return true
    }

    // Case 3: Target is in playNext section
    if index >= playNextStart && index < playNextEnd {
      let targetTrack = actualQueue[index]
      // Remove all playNext tracks before the target (by ID lookup, not position)
      if let targetIdx = playNextStack.firstIndex(where: { $0.id == targetTrack.id }), targetIdx > 0 {
        // Remove tracks before target, but keep the currently playing track
        // (rebuildAVQueueFromCurrentPosition will skip it by ID)
        playNextStack.removeSubrange(0..<targetIdx)
      }
      if let player = player, currentRepeatMode == .track {
        dropRepeatCopyBeforeManualSkip(player)
      } else {
        rebuildAVQueueFromCurrentPosition()
      }
      player?.advanceToNextItem()
      return true
    }

    // Case 4: Target is in upNext section
    if index >= upNextStart && index < upNextEnd {
      let targetTrack = actualQueue[index]
      playNextStack.removeAll()
      // Remove all upNext tracks before the target (by ID lookup, not position)
      if let targetIdx = upNextQueue.firstIndex(where: { $0.id == targetTrack.id }), targetIdx > 0 {
        upNextQueue.removeSubrange(0..<targetIdx)
      }
      if let player = player, currentRepeatMode == .track {
        dropRepeatCopyBeforeManualSkip(player)
      } else {
        rebuildAVQueueFromCurrentPosition()
      }
      player?.advanceToNextItem()
      return true
    }

    // Case 5: Target is in remaining original tracks
    if index >= originalRemainingStart {
      let targetTrack = actualQueue[index]
      guard let originalIndex = currentTracks.firstIndex(where: { $0.id == targetTrack.id }) else { return false }

      playNextStack.removeAll()
      upNextQueue.removeAll()
      currentTemporaryType = .none

      let result = rebuildQueueFromPlaylistIndex(index: originalIndex)
      checkUpcomingTracksForUrls(lookahead: lookaheadCount)
      return result
    }

    checkUpcomingTracksForUrls(lookahead: lookaheadCount)
    return false
  }

  func playFromIndexInternal(index: Int) {
    playNextStack.removeAll()
    upNextQueue.removeAll()
    currentTemporaryType = .none
    _ = rebuildQueueFromPlaylistIndex(index: index)
  }

  func determineCurrentTemporaryType() -> TemporaryType {
    guard let trackId = activeCurrentTrackId else { return .none }
    if playNextStack.contains(where: { $0.id == trackId }) { return .playNext }
    if upNextQueue.contains(where: { $0.id == trackId }) { return .upNext }
    return .none
  }
}

File: react-native-nitro-player/android/src/main/java/com/margelo/nitro/nitroplayer/core/TrackPlayerQueue.kt
@file:Suppress("ktlint:standard:max-line-length")

package com.margelo.nitro.nitroplayer.core

import androidx.media3.common.Player
import com.margelo.nitro.nitroplayer.CurrentPlayingType
import com.margelo.nitro.nitroplayer.PlayerState
import com.margelo.nitro.nitroplayer.TrackItem
import com.margelo.nitro.nitroplayer.TrackPlayerState
import com.margelo.nitro.nitroplayer.Variant_NullType_String
import com.margelo.nitro.nitroplayer.Variant_NullType_TrackItem

/**
 * State read + index-navigation — all public functions are suspend and run on the
 * player thread via withPlayerContext.
 */

// ── Player state ──────────────────────────────────────────────────────────

suspend fun TrackPlayerCore.getState(): PlayerState = withPlayerContext { getStateInternal() }

internal fun TrackPlayerCore.getStateInternal(): PlayerState {
    if (!isExoInitialized) {
        return PlayerState(
            currentTrack = null,
            currentPosition = 0.0,
            totalDuration = 0.0,
            currentState = TrackPlayerState.STOPPED,
            currentPlaylistId = currentPlaylistId?.let { Variant_NullType_String.create(it) },
            currentIndex = -1.0,
            currentPlayingType = CurrentPlayingType.NOT_PLAYING,
        )
    }
    val track = getCurrentTrack()
    val currentTrack: Variant_NullType_TrackItem? = track?.let { Variant_NullType_TrackItem.create(it) }
    val position = exo.currentPosition / 1000.0
    val duration = if (exo.duration > 0) exo.duration / 1000.0 else 0.0
    val state =
        when (exo.playbackState) {
            Player.STATE_IDLE -> TrackPlayerState.STOPPED
            Player.STATE_BUFFERING -> if (exo.playWhenReady) TrackPlayerState.PLAYING else TrackPlayerState.PAUSED
            Player.STATE_READY -> if (exo.isPlaying) TrackPlayerState.PLAYING else TrackPlayerState.PAUSED
            Player.STATE_ENDED -> TrackPlayerState.STOPPED
            else -> TrackPlayerState.STOPPED
        }
    val playingType =
        if (track == null) {
            CurrentPlayingType.NOT_PLAYING
        } else {
            when (currentTemporaryType) {
                TrackPlayerCore.TemporaryType.NONE -> CurrentPlayingType.PLAYLIST
                TrackPlayerCore.TemporaryType.PLAY_NEXT -> CurrentPlayingType.PLAY_NEXT
                TrackPlayerCore.TemporaryType.UP_NEXT -> CurrentPlayingType.UP_NEXT
            }
        }
    return PlayerState(
        currentTrack = currentTrack,
        currentPosition = position,
        totalDuration = duration,
        currentState = state,
        currentPlaylistId = currentPlaylistId?.let { Variant_NullType_String.create(it) },
        currentIndex = if (exo.currentMediaItemIndex >= 0) exo.currentMediaItemIndex.toDouble() else -1.0,
        currentPlayingType = playingType,
    )
}

// ── Actual queue ──────────────────────────────────────────────────────────

suspend fun TrackPlayerCore.getActualQueue(): List<TrackItem> = withPlayerContext { getActualQueueInternal() }

internal fun TrackPlayerCore.getActualQueueInternal(): List<TrackItem> {
    if (!isExoInitialized) return emptyList()
    val currentIndex = currentTrackIndex
    if (currentIndex < 0) return emptyList()

    val queue = ArrayList<TrackItem>(currentTracks.size + playNextStack.size + upNextQueue.size)

    // Tracks before current (include currentTrackIndex when a temp track is playing)
    val beforeEnd =
        if (currentTemporaryType != TrackPlayerCore.TemporaryType.NONE) {
            minOf(currentIndex + 1, currentTracks.size)
        } else {
            currentIndex
        }
    if (beforeEnd > 0) queue.addAll(currentTracks.subList(0, beforeEnd))

    // Current track
    getCurrentTrack()?.let { queue.add(it) }

    val currentId = exo.currentMediaItem?.mediaId?.let { extractTrackId(it) }

    // playNext — skip the currently playing track by ID (not position)
    if (currentTemporaryType == TrackPlayerCore.TemporaryType.PLAY_NEXT && currentId != null) {
        var skipped = false
        for (track in playNextStack) {
            if (!skipped && track.id == currentId) {
                skipped = true
                continue
            }
            queue.add(track)
        }
    } else if (currentTemporaryType != TrackPlayerCore.TemporaryType.PLAY_NEXT) {
        queue.addAll(playNextStack)
    }

    // upNext — skip the currently playing track by ID (not position)
    if (currentTemporaryType == TrackPlayerCore.TemporaryType.UP_NEXT && currentId != null) {
        var skipped = false
        for (track in upNextQueue) {
            if (!skipped && track.id == currentId) {
                skipped = true
                continue
            }
            queue.add(track)
        }
    } else if (currentTemporaryType != TrackPlayerCore.TemporaryType.UP_NEXT) {
        queue.addAll(upNextQueue)
    }

    // Remaining original tracks
    if (currentIndex + 1 < currentTracks.size) {
        queue.addAll(currentTracks.subList(currentIndex + 1, currentTracks.size))
    }
    return queue
}

// ── Index navigation ──────────────────────────────────────────────────────

suspend fun TrackPlayerCore.skipToIndex(index: Int): Boolean = withPlayerContext { skipToIndexInternal(index) }

private fun TrackPlayerCore.skipToIndexInternal(index: Int): Boolean {
    if (!isExoInitialized) return false
    val actualQueue = getActualQueueInternal()
    if (index < 0 || index >= actualQueue.size) return false

    val currentPos = if (currentTemporaryType != TrackPlayerCore.TemporaryType.NONE) currentTrackIndex + 1 else currentTrackIndex
    val effectivePlayNextSize = if (currentTemporaryType == TrackPlayerCore.TemporaryType.PLAY_NEXT) maxOf(0, playNextStack.size - 1) else playNextStack.size
    val effectiveUpNextSize = if (currentTemporaryType == TrackPlayerCore.TemporaryType.UP_NEXT) maxOf(0, upNextQueue.size - 1) else upNextQueue.size

    val playNextStart = currentPos + 1
    val playNextEnd = playNextStart + effectivePlayNextSize
    val upNextStart = playNextEnd
    val upNextEnd = upNextStart + effectiveUpNextSize
    val originalRemainingStart = upNextEnd

    if (index < currentPos) {
        playFromIndexInternal(index)
        return true
    }
    if (index == currentPos) {
        exo.seekTo(0)
        return true
    }

    if (index in playNextStart until playNextEnd) {
        val targetTrack = actualQueue[index]
        // Remove all playNext tracks before the target (by ID lookup, not position)
        val targetIdx = playNextStack.indexOfFirst { it.id == targetTrack.id }
        if (targetIdx > 0) playNextStack.subList(0, targetIdx).clear()
        rebuildQueueFromCurrentPosition()
        exo.seekToNext()
        return true
    }

    if (index in upNextStart until upNextEnd) {
        val targetTrack = actualQueue[index]
        playNextStack.clear()
        // Remove all upNext tracks before the target (by ID lookup, not position)
        val targetIdx = upNextQueue.indexOfFirst { it.id == targetTrack.id }
        if (targetIdx > 0) upNextQueue.subList(0, targetIdx).clear()
        rebuildQueueFromCurrentPosition()
        exo.seekToNext()
        return true
    }

    if (index >= originalRemainingStart) {
        val targetTrack = actualQueue[index]
        val originalIndex = currentTracks.indexOfFirst { it.id == targetTrack.id }
        if (originalIndex == -1) return false
        playNextStack.clear()
        upNextQueue.clear()
        currentTemporaryType = TrackPlayerCore.TemporaryType.NONE
        rebuildQueueAndPlayFromIndex(originalIndex)
        checkUpcomingTracksForUrls(lookaheadCount)
        return true
    }

    checkUpcomingTracksForUrls(lookaheadCount)
    return false
}

suspend fun TrackPlayerCore.playFromIndex(index: Int) = withPlayerContext { playFromIndexInternal(index) }

internal fun TrackPlayerCore.playFromIndexInternal(index: Int) {
    playNextStack.clear()
    upNextQueue.clear()
    currentTemporaryType = TrackPlayerCore.TemporaryType.NONE
    rebuildQueueAndPlayFromIndex(index)
}
